docs: service-registry-doc (#268)
axelarnetwork/axelar-amplifier: doc/src/contracts/service_registry.md

# Service Registry

> The service registry keeps track of the pool of workers that are related to each service.

Reviewer: ```suggestion
The service registry keeps track of the pool of workers that vote and sign for each chain.
```

The core functionalities, such as registering a new service, worker authorization and un-authorization can only be called 

> from a governance address. To become active and be eligible to participate in voting for a specific service, 

Reviewer: ```suggestion
from a governance address. Worker bonding and unbonding, as well as registering support for specific chains, are called by the worker themselves. To become active and be eligible to participate in voting for a specific chain, 
```

> each worker should first be authorized and the bond enough stake. Service registry is used by ampd, verifier, and multisig prover.

Reviewer: ```suggestion
each worker should first be authorized and bond enough stake, and then register support for that chain. 

Service registry is used by ampd, verifier, and multisig prover.
```

> 

Reviewer: Can we add a paragraph here talking about what exactly is a service? Versus a chain. Call out specifically that we plan to use one service, and all chains are within that service.

Author: added

    UnauthorizeWorkers {
        workers: Vec<String>,
        service_name: String,
    },

    // Register support for the specified chains. Called by the worker.
    RegisterChainSupport {
        service_name: String,
        chains: Vec<ChainName>,
    },
    // Deregister support for the specified chains. Called by the worker.
    DeregisterChainSupport {
        service_name: String,
        chains: Vec<ChainName>,
    },

    // Locks up any funds sent with the message as stake. Called by the worker.
    BondWorker {
        service_name: String,
    },
    // Initiates unbonding of staked funds. Called by the worker.
    UnbondWorker {
        service_name: String,
    },
    // Claim previously staked funds that have finished unbonding. Called by the worker.
    ClaimStake {
        service_name: String,
    },
}

```

## Service Registry graph
```mermaid
flowchart TD
subgraph Axelar
    Vr{"Voting Verifier"}
    Pr{"Prover"}
    R{"Service Registry"}
end
OC{"Workers"}

Vr -- "GetActiveWorkers" --> R
Pr -- "GetActiveWorkers" --> R
R -- "Vec[Worker]" --> Vr

> R -- "Vec[Worker]" --> Pr

Reviewer: I think for clarity we can remove these lines

> OC -- "Vote(poll_id, votes)" --> Vr

Reviewer: Remove, I don't think this is necessary

OC -- "De/RegisterChainSupport" --> R
OC -- "Un/BondWorker" --> R
OC -- "ClaimStake" --> R
```

## Service Registry sequence diagram

```mermaid
sequenceDiagram
autonumber
box LightYellow Axelar
    participant Service Registry
end
actor Governance
actor Worker

Governance->>+Service Registry: Register Service
Governance->>+Service Registry: Authorize Workers

Worker->>+Service Registry: Register Chain Support

> Worker->>+Service Registry: Bond Worker

Reviewer: Can we swap the order of these?

Author: done

Worker->>+Service Registry: Claim Stake

> Service Registry->>Worker: returns BankMsg with amount set to released_bond

Reviewer: Let's remove the claiming stake part here. I'd rather the sequence diagram just show the worker onboarding process. Claiming stake is off boarding. Also, the worker actually needs to unbond first, before claiming stake. If you want, we can add a separate diagram for off boarding. Claiming stake requires waiting for some unbonding period.

Author: I removed the lines

```

1. The Governance registers a new service by providing the necessary parameters for the service.

> 2. Governance is also responsible for authorizing workers to join the service by sending an `Authorize Workers` message.

Reviewer: Can we add something saying authorizing and bonding can be done in any order?

3. Workers register support for specific chains within the service by specifying service name and chain names.
4. Workers bond to the service, providing stake, by sending a `Bond Worker` message with appropriate funds denominator.
5. Workers can claim released stake with `Claim Stake` message, and the Service Registry responds with a BankMsg containing the released bond.

> 

Reviewer: We probably also want to add something that says workers need to register their public key to participate in signing, and add a link to the multisig contract documentation.

Author: I added a Notes part

# Service Registry

The service registry keeps track of the pool of workers that vote and sign for each chain.
The core functionalities, such as registering a new service, worker authorization and un-authorization can only be called 
from a governance address. Worker bonding and unbonding, as well as registering support for specific chains, are called by the worker themselves. 
To become active and be eligible to participate in voting for a specific chain,
each worker should first be authorized and bond enough stake, and then register support for that chain.
Service registry is used by ampd, verifier, and multisig prover.

The term `service` refers to an upper-level entity that includes several 
`chains`. The difference is in how they are related to each other, which is 
hierarchical; a service is like an umbrella that regulates the activities of 
several chains that fall under its purview. The service defines common 
parameters, such as worker requirements, bonding details, and unbonding periods, which are applicable to all associated chains. 

> Thus, we use a single instance of service registry to organize and coordinate activities across all chains.

Reviewer: Nice summary!
